File: java/parsers/src/main/java/jp/co/teruuu/parser/common/ParseResult.java
package jp.co.teruuu.parser.common;

import java.util.function.*;
public interface ParseResult<T> {
    public class Success<T> implements ParseResult<T> {
        public final T value;
        public final String next;
        public Success(T value, String next) {
            this.value = value;
            this.next = next;
        }

        @Override
        public <U> ParseResult<U> map(Function<T, U> fn) {
            return new Success<U>(fn.apply(value), next);
        }

        @Override
        public String toString() {
            return "Success(" + value + ", " + next + ")";
        }
    }

    public class Failure<T> implements ParseResult<T> {
        public final String message;
        public final String next;

        public Failure(String message, String next) {
            this.message = message;
            this.next = next;
        }

        @Override
        public <U> ParseResult<U> map(Function<T, U> fn) {
            return (ParseResult<U>)this;
        }

        @Override
        public String toString() {
            return "Failure(" + message + ", " + next;
        }
    }

    <U> ParseResult<U> map(Function<T, U> fn);
}

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/ParseResult.scala
package jp.co.teruuu.parser.common

trait ParseResult[T] {
}

case class ParseOk[T](value: T, next: String) extends ParseResult[T] {
  override def toString(): String = "Success(" + value + ", " + next + ")"
}

case class ParseNg[T](message: String, next: String) extends ParseResult[T] {
  override def toString(): String = "Failure(" + message + ", " + next + ")"
}
